### PR TITLE
Opro implementation

### PR DESCRIPTION
**Pull request by finitearth**

### CODE REVIEW

File: promptolution/optimizers/opro.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            self.meta_prompt = "".join(f.readlines())
+        
+        super().__init__(**args)
+        self.meta_prompt = self.meta_prompt.replace("<task_description>", self.task.description)
+
+        self.scores = [self.task.evaluate(p, self.predictor) for p in self.prompts]
+
+    def _sample_examples(self):
+        sample_x = np.random.choice(self.task.xs, self.n_samples)
+        sample_y = np.random.choice(self.task.ys, self.n_samples)
```
Comment (timo282) [inline promptolution/optimizers/opro.py:29]: better sample the ideas --> x's and y's may not match

File: promptolution/optimizers/opro.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import numpy as np
+
+from promptolution.llms.base_llm import BaseLLM
+from promptolution.optimizers.base_optimizer import BaseOptimizer
+
+
+class Opro(BaseOptimizer):
+    """
+    Opro: Optimization by PROmpting
+
```
Comment (mo374z) [inline promptolution/optimizers/opro.py:12]: lets describe the idea at least high level s.t. users does not have to skim the paper if not familiar with the base idea

File: promptolution/optimizers/opro.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    Proposed by the paper "Large Language Models as Optimizers" by Yang et. al: https://arxiv.org/abs/2309.03409
+
+    """
+
+    def __init__(self, llm: BaseLLM, n_samples: int = 2, **args):
```
Comment (mo374z) [inline promptolution/optimizers/opro.py:17]: specify what n_samples means (we use this also as n_samples of prompts) our users should not get confused by this

File: promptolution/optimizers/opro.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        sample_y = self.task.ys[idx]
+
+        return "\n".join([f"Input: {x}\nOutput: {y}" for x, y in zip(sample_x, sample_y)])
+
+    def _format_old_instructions(self):
+        return "".join(
+            [f"Old instruction: {prompt}\nScore: {score}\n\n" for prompt, score in zip(self.prompts, self.scores)]
+        )
+
+    def optimize(self, n_steps: int) -> List[str]:
```
Comment (mo374z) [inline promptolution/optimizers/opro.py:40]: add doc comment

File: promptolution/optimizers/opro.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        self.scores = [self.task.evaluate(p, self.predictor) for p in self.prompts]
+
+    def _sample_examples(self):
+        idx = np.random.choice(len(self.task.xs), self.n_samples)
+        sample_x = self.task.xs[idx]
+        sample_y = self.task.ys[idx]
+
+        return "\n".join([f"Input: {x}\nOutput: {y}" for x, y in zip(sample_x, sample_y)])
+
+    def _format_old_instructions(self):
```
Comment (mo374z) [inline promptolution/optimizers/opro.py:35]: add doc comment

File: promptolution/optimizers/opro.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        self.n_samples = n_samples
+        with open("templates/opro_template.txt") as f:
+            self.meta_prompt = "".join(f.readlines())
+
+        super().__init__(**args)
+        self.meta_prompt = self.meta_prompt.replace("<task_description>", self.task.description)
+
+        self.scores = [self.task.evaluate(p, self.predictor) for p in self.prompts]
+
+    def _sample_examples(self):
```
Comment (mo374z) [inline promptolution/optimizers/opro.py:28]: add doc comment

File: promptolution/optimizers/opro.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    """
+    Opro: Optimization by PROmpting
+
+    Proposed by the paper "Large Language Models as Optimizers" by Yang et. al: https://arxiv.org/abs/2309.03409
+
+    """
+
+    def __init__(self, llm: BaseLLM, n_samples: int = 2, **args):
+        self.llm = llm
+        self.n_samples = n_samples
```
Comment (mo374z) [inline promptolution/optimizers/opro.py:19]: check whether n_samples is in the valid value range (>=0 i guess)

**Review by mo374z — APPROVED** [review]

**Review by timo282 — APPROVED** [review]

**Pull request merged**